File: sem4/mpp/mpp-proiect-csharp-vreauladudu/server/MotoServer.cs
using mpp_proiect_csharp_vreauladudu.repository;
using mpp_proiect_csharp_vreauladudu.service;
using mpp_proiect_csharp_vreauladudu.domain;
using System.Data.SQLite;
using services;

namespace Server.Impl
{
    public class MotoServer : IMotoService
    {
        private MotorcycleRepository motorcycles;
        private ParticipantRepository participants;
        private UserRepository users;
        private readonly IDictionary<int, IMotoObserver> loggedClients;

        public MotoServer(MotorcycleRepository motorcycles, ParticipantRepository participants, UserRepository users)
        {
            this.motorcycles = motorcycles;
            this.participants = participants;
            this.users = users;
            this.loggedClients = new Dictionary<int, IMotoObserver>();
        }

        public void login(User user, IMotoObserver client)
        {
            try
            {
                User userToLogIn = this.users.isValidUser(user.Username, user.Password);
                if (userToLogIn == null)
                {
                    throw new MotoException("Authentication failed.");
                }

                if (this.loggedClients.ContainsKey(userToLogIn.Id))
                {
                    throw new MotoException("User already logged in.");
                }

                this.loggedClients[userToLogIn.Id] = client;
            }
            catch (SQLiteException exception)
            {
                Console.Write(exception.StackTrace);
            }
        }

        public void logout(User user, IMotoObserver client)
        {
            try
            {
                User userToLogOut = this.users.isValidUser(user.Username, user.Password);
                if (userToLogOut == null)
                {
                    throw new MotoException("Unexpected error.");
                }

                if (!this.loggedClients.ContainsKey(userToLogOut.Id))
                {
                    throw new MotoException("User " + user + " is not logged in.");
                }

                this.loggedClients.Remove(userToLogOut.Id);
            }
            catch (SQLiteException exception)
            {
                Console.Write(exception.StackTrace);
            }
        }

        public Dictionary<int, List<Motorcycle>> listRaces(IMotoObserver client)
        {
            IEnumerable<Motorcycle> motorcycles = this.motorcycles.findAll();
            return motorcycles.GroupBy(motorcycle => motorcycle.Capacity).ToDictionary(motorcycle => motorcycle.Key, motorcycle => motorcycle.Select(r => r).ToList());
        }

        public IEnumerable<Motorcycle> listMotorcycles(IMotoObserver client)
        {
            return this.motorcycles.findAll();
        }

        public List<Motorcycle> listMotorcyclesByCapacity(int capacity, IMotoObserver client)
        {
            return this.motorcycles.findByCapacity(capacity);
        }

        public List<Participant> findParticipantForRace(int motorcycleId, IMotoObserver client)
        {
            return this.participants.findByMotorcycleId(motorcycleId);
        }

        public List<Motorcycle> listMotorcyclesByBrand(String brand, IMotoObserver client)
        {
            return this.motorcycles.findByBrand(brand);
        }

        public Motorcycle findMotorcycleById(int motorcycleId, IMotoObserver client)
        {
            return this.motorcycles.findById(motorcycleId);
        }

        public void addParticipant(String participantName, String motorcycleBrand, int motorcycleCapacity, IMotoObserver client)
        {
            int maximumId = 0;
            List<Participant> participantList = this.participants.findAll().ToList();
            foreach (Participant participant in participantList)
            {
                if (participant.Id > maximumId)
                {
                    maximumId = participant.Id;
                }
            }

            List<Motorcycle> motorcycleListByBrand = this.listMotorcyclesByBrand(motorcycleBrand, client);
            List<Motorcycle> motorcycleListByCapacity = this.listMotorcyclesByCapacity(motorcycleCapacity, client);
            try
            {
                Motorcycle motorcycle = motorcycleListByBrand.Distinct().Where(motorcycleItem => motorcycleListByCapacity.Contains(motorcycleItem)).First();
                Participant participant = new Participant();
                participant.Id = maximumId + 1;
                participant.Name = participantName;
                participant.MotorcycleID = motorcycle.Id;
                this.participants.save(participant);
                notifyOthersParticipantAdded(participant);
            }
            catch (Exception exception)
            {
                Console.Write(exception.StackTrace);
            }
        }

        private void notifyOthersParticipantAdded(Participant participant)
        {
            IEnumerable<User> users = this.users.findAll();
            foreach (User user in users)
            {
                if (this.loggedClients.ContainsKey(user.Id))
                {
                    IMotoObserver motoClient = this.loggedClients[user.Id];
                    Task.Run(() => motoClient.participantAdded(participant));
                }
            }
        }
    }
}

File: sem4/mpp/mpp-proiect-java-vreauladudu/server/src/main/java/moto/server/MotoServer.java
package moto.server;

import moto.Motorcycle;
import moto.Participant;
import moto.User;
import moto.repository.MotorcycleRepository;
import moto.repository.ParticipantRepository;
import moto.repository.UserRepository;
import moto.services.IMotoObserver;
import moto.services.IMotoService;
import moto.services.MotoException;

import java.sql.SQLException;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class MotoServer implements IMotoService
{
    private MotorcycleRepository motorcycles;
    private ParticipantRepository participants;
    private UserRepository users;
    private Map<Integer, IMotoObserver> loggedClients;

    public MotoServer(MotorcycleRepository motorcycles, ParticipantRepository participants, UserRepository users)
    {
        this.motorcycles = motorcycles;
        this.participants = participants;
        this.users = users;
        this.loggedClients = new ConcurrentHashMap<>();
    }

    public synchronized void login(User user, IMotoObserver client) throws MotoException
    {
        try
        {
            User userToLogIn = this.users.isValidUser(user.getUsername(), user.getPassword());
            if (userToLogIn == null)
            {
                throw new MotoException("Authentication failed.");
            }

            if (this.loggedClients.get(userToLogIn.getId()) != null)
            {
                throw new MotoException("User already logged in.");
            }

            this.loggedClients.put(userToLogIn.getId(), client);
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
    }

    public synchronized void logout(User user, IMotoObserver client) throws MotoException
    {
        try
        {
            User userToLogOut = this.users.isValidUser(user.getUsername(), user.getPassword());
            if (userToLogOut == null)
            {
                throw new MotoException("Unexpected error.");
            }

            if (this.loggedClients.get(userToLogOut.getId()) == null)
            {
                throw new MotoException("User " + user + " is not logged in.");
            }

            this.loggedClients.remove(userToLogOut.getId());
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
    }

    public synchronized Map<Integer, List<Motorcycle>> listRaces(IMotoObserver client)
    {
        Iterable<Motorcycle> motorcycles = this.motorcycles.findAll();
        return StreamSupport.stream(motorcycles.spliterator(), false).collect(Collectors.groupingBy(Motorcycle::getCapacity));
    }

    public synchronized Iterable<Motorcycle> listMotorcycles(IMotoObserver client)
    {
        return this.motorcycles.findAll();
    }

    public synchronized List<Motorcycle> listMotorcyclesByCapacity(Integer capacity, IMotoObserver client)
    {
        try
        {
            return this.motorcycles.findByCapacity(capacity);
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
        return null;
    }

    public synchronized List<Participant> findParticipantForRace(Integer motorcycleId, IMotoObserver client)
    {
        try
        {
            return this.participants.findByMotorcycleId(motorcycleId);
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
        return null;
    }

    public synchronized List<Motorcycle> listMotorcyclesByBrand(String brand, IMotoObserver client)
    {
        try
        {
            return this.motorcycles.findByBrand(brand);
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
        return null;
    }

    public synchronized Motorcycle findMotorcycleById(Integer motorcycleId, IMotoObserver client)
    {
        try
        {
            return this.motorcycles.findById(motorcycleId);
        }
        catch (SQLException exception)
        {
            exception.printStackTrace();
        }
        return null;
    }

    public synchronized void addParticipant(String participantName, String motorcycleBrand, Integer motorcycleCapacity, IMotoObserver client)
    {
        Integer maximumId = 0;
        Iterable<Participant> participantList = this.participants.findAll();
        for (Participant participant : participantList)
        {
            if (participant.getId() > maximumId)
            {
                maximumId = participant.getId();
            }
        }

        List<Motorcycle> motorcycleListByBrand = listMotorcyclesByBrand(motorcycleBrand, client);
        List<Motorcycle> motorcycleListByCapacity = listMotorcyclesByCapacity(motorcycleCapacity, client);
        Optional<Motorcycle> motorcycle = motorcycleListByBrand.stream().distinct().filter(motorcycleListByCapacity::contains).findFirst();
        if (motorcycle.isPresent())
        {
            Participant participant = new Participant();
            participant.setId(maximumId + 1);
            participant.setName(participantName);
            participant.setMotorcycleID(motorcycle.get().getId());
            this.participants.save(participant);
            this.notifyOthersParticipantAdded(participant);
        }
    }

    private synchronized void notifyOthersParticipantAdded(Participant participant)
    {
        ExecutorService executor = Executors.newFixedThreadPool(4);

        this.users.findAll().forEach(user ->
        {
            if (this.loggedClients.get(user.getId()) != null)
            {
                IMotoObserver motoClient = this.loggedClients.get(user.getId());
                executor.execute(() ->
                {
                    try
                    {
                        motoClient.participantAdded(participant);
                    }
                    catch (MotoException exception)
                    {
                        System.err.println("Moto exception in thread " + exception.getMessage());
                        exception.printStackTrace();
                    }
                });
            }
        });

        executor.shutdown();
    }
}